### ISSUE TITLE
Submit not fired when nested inside another Element with click binding

### ISSUE THREAD

**Issue by stewieoO** [angular/angular#25403]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
When a `(submit)` binding is somewhere nested inside another element that has a `(click)` binding and that binding returns false (implicitly or explicitly), the submit will not be called.

Nested `(click)` bindings, however, will be called as expected.

## Expected behavior
A nested `(submit)` binding should be called even when a parent element has a `(click)` binding that returns false.

## Minimal reproduction of the problem with instructions
Create a div that has a `(click)` binding where the method or logic returns `false`.
Nest a form inside that div which has a submit button and add the `(submit)` binding.

[Stackblitz Reproduction](https://stackblitz.com/edit/angular-adefxu)

You can see in the `AppComponent.ngfactory.js` where the click binding happens in this piece of code
```
if (('click' === en)) {
  _co.alert('outer');
  var pd_0 = (false !== false);
  ad = (pd_0 && ad);
}
return ad;
```
when the last line returns false, the issue occurs.

## What is the motivation / use case for changing the behavior?

We were assigning a variable in the html template:
`<div (click)="myvar = false;"></div>`

## Environment

<pre><code>
Angular version: 6.1.2

  
Browser:
- [x] Chrome (desktop) version 68
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [x] Firefox version 61
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [x] IE version 11
- [x] Edge version 41
 
For Tooling issues:
- Node version: 8.11.1
- Platform:  Windows

Others:

</code></pre>

**Comment by mhevery** [comment]
Returning false from the handler cancels the event. So from my point of view it works as intended.

**Comment by stewieoO** [comment]
Should'nt it behave the same with click+click and click+submit though?

**Comment by mhevery** [comment]
Could you give me exact steps on where I should click and what you expect should happen? Your demo has lots of buttons and it is not clear to me what you think should happen on each button, and I don't have the time to reverse engineer your intent.

**Comment by stewieoO** [comment]
Sure. I thought with the explanation in the issue it would become clear. My bad.

In each part you have two elements where one is nested into the other, hence the name "Inner" and "Outer". Each click listener emits an alert. The "Outer" element's listener also returns either true, false or null.

In the first part with the headline "2 Clicks" it is simply two divs, both with a click listener.
If you click the "Inner" it will show the "Inner" alert and then the "outer" alert, as expected.
The "Outer" listener returns false.

In the second part the inner elements are forms and don't have a click listener, but a submit listener.
This part also works as expected and the outer listeners return null and/or true.
You click the "Submit", and both listeners (inner, outer) are being called.

In the third part the "Outer" listener returns false and by clicking the "Submit" only the "Outer" listener is called.

Also 

> Returning false from the handler cancels the event. So from my point of view it works as intended.

When i let the "Inner" listener return "false" in the first part, both listeners are called.
I've updated the [example ](https://stackblitz.com/edit/angular-qm6zyh) with a bit more explanation.

I would expect each case to call both listeners when i click the "Inner" element.

**Comment by pkozlowski-opensource** [comment]
tl;dr; things work as designed and the confusion is about what `onclick="alert('outter'); false"` does.

For longer explanation we need to observe that: 
* a single act of clicking on an inner button will result in 2 DOM events: `click` and `submit`;
* a browser will first dispatch and bubble `click` event, _than_ will dispatch `submit` event;
* preventing default on the `click` event will prevent `submit` event from being dispatched.

Now let's look at the simplified stackblitz where we can see that _similar_ syntax works differently in pure JS and Angular: https://stackblitz.com/edit/angular-87pgkq?file=src/app/app.component.html

This issue claims that both should behave the same but this claim is only legit if we _assume_ that `onclick="alert('outter'); false"` and `(click)="alert('outer'); false"` _behave_ the same way. This turns out _not_ to be true - while the syntax is deceptively similar the actual semantics are not!

Browser / JS takes event handler statements (`alert('outer'); false`) and wraps into a function so the actual registered event handler looks like so:

```
function(event) {
  alert('outer'); 
  false
}
```

Notice that here we've got a "dangling" `false` without any return value so this handler function always returns `undefined`, regardless of what you put in there (this is why it works with `false`, `true`.

Angular is more elaborate in the way it generates event handler functions and the actual code will be:

```javascript
function(_v, en, $event) {
            var ad = true;
            var _co = _v.component;
            if (('click' === en)) {
                _co.alert('outer');
                var pd_0 = (false !== false);
                ad = (pd_0 && ad);
            }
            return ad;
        }
```

What Angular does is that it _parses_  `alert('outer'); false` and _assumes_ that the last statement is the _return value_ of the event handler.

This explain the difference: in the in-line JavaScript handler you have to insert `return` keyword for you while Angular does it for you. In other words passing `alert('outer'); false` to JS and Angular will "generate" 2 _slightly_ different event handler functions. 

In JS:
```javascript
function(event) {
  alert('outer'); 
  false
}
```

In Angular (simplified):

```javascript
function(event) {
  alert('outer'); 
  return false;
}
```

At the end of a day this all boils down to the following question: should Angular add the `return` keyword in the generated event handler code? I believe so, as in-line event handlers are just too confusing (as your example proves!).

I'm going to close this one as it is working as designed. I fully acknowledge that there might be confusion here with in-line event handlers but I believe that Angular approach is more predictable.

**Comment by pkozlowski-opensource** [comment]
BTW, if we take different semantics into account you can see that both regular HTML+JS and Angular behaviours are the same.
